### ISSUE TITLE
[HOLD for payment 2023-05-16] Add custom renderer for user mentions

### ISSUE THREAD

**Issue by puneetlath** [Expensify/App#17885]
As part of the [mentions project](https://docs.google.com/document/d/1vhvpwQDStA0lcsI0Z33pXOKPM6nXGRuplvQTFBWvCyA/edit#heading=h.ns12dr1f35rx), we will be adding a [custom renderer](https://meliorence.github.io/react-native-render-html/docs/guides/custom-renderers#component-based-custom-rendering) called MentionUserRenderer for the <mention-user> tag. The custom renderer will be similar to the [AnchorRenderer](https://github.com/Expensify/App/blob/28bda4de991d8a6579e0549ee250bde66c928265/src/components/HTMLEngineProvider/HTMLRenderers/AnchorRenderer.js). 

The details of the implementation are described in the doc [here](https://docs.google.com/document/d/1vhvpwQDStA0lcsI0Z33pXOKPM6nXGRuplvQTFBWvCyA/edit#heading=h.5lyudapet5j6).

**Comment by RobertKozik** [comment]
Hi 👋🏼  I'm from Software Mansion and I'll handle this task

**Comment by MelvinBot** [comment]
📣 @RobertKozik! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/231263719-a8a3b7d3-36a0-4d7f-8ad4-cebe1f560fa4.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by puneetlath** [comment]
Ok great @robertKozik. I have assigned you to the issue and shared the design doc with you. Please feel free to hit me up here or in slack if you have any questions!

**Comment by puneetlath** [comment]
@robertKozik if you want to test this, the easiest way will be to: 

1. Add the changes from [this PR](https://github.com/Expensify/expensify-common/pull/524) in `App/node_modules/expensify-common/lib/ExpensiMark.js`. That will make it so that when you type `@here` or `@user@domain.com` it gets the <mention-here> or <mention-user> html tags added. 
2. Force the app offline. This way when you send a message it'll just show the optimistic message, not the one returned by the server. This is because the server will strip any html tags it doesn't recognize (we just merged a PR to allow these two, so it won't be live for another few days)

<img width="364" alt="2023-04-27_21-17-33" src="https://user-images.githubusercontent.com/1371996/235031444-8a202aa9-88b8-484e-82f6-fe6f19757fbf.png">

**Comment by robertKozik** [comment]
Thats way more convenient than what I used to do (patching the `ReportActionItemFragment` to replace all occurances). Thanks!

**Comment by robertKozik** [comment]
Hey Hey, two questions:
1. In the mockups in design doc the tooltip which turn on after hovering the user mention includes user profile image. The snippet of userMentionRenderer does not include it. Is it still in the scope for the `Before EC3` version? 
2. User mentions which comes from the ExpensiMark does not include fields like login, it's only: `<mention-user>@user@mail</mention-user>` Is it intended and the login handle is only provided inside the body of the HTML tag?

**Comment by puneetlath** [comment]
Great questions. 

1. Given that we're using primary logins, I think we can skip the tooltip altogether for the before EC3 portion of mentions. I don't think any tooltip is needed. We just need to make sure that the profile page is opened when you click on a mention. 

![image](https://user-images.githubusercontent.com/1371996/235157749-d10def9d-79ee-46f0-a06a-d01f2bef6ab1.png)

2. Yes, for now there won't be a login attribute. There won't be any html attributes. Just the text between the tags.

**Comment by robertKozik** [comment]
Thanks for answers, today I should create the PR with it.

**Comment by puneetlath** [comment]
Awesome!

**Comment by robertKozik** [comment]
Hi! Sadly I didn't finish it :/ I've encountered two problems:
1. When I've patched the `ExpensiMark` file, on `Safari` browser app is not loading for me and gets following error. Looks like safari can't handle new regex, but I'm not sure about it.
<img width="497" alt="image" src="https://user-images.githubusercontent.com/61146594/235242637-976d51e5-1f0b-49e4-870a-a8b4519c41f9.png">

2. Android native app can't render mention when it's wrapped in Pressable component. Only after refreshing the code, mentions are appearing 
<img width="359" alt="image" src="https://user-images.githubusercontent.com/61146594/235251448-eafea8e3-4f0f-46d7-b8fd-6fe6c5606484.png">

I'll investigate it further

**Comment by puneetlath** [comment]
Hmm, interesting. It looks like our [custom renderer for Anchor tags](https://github.com/Expensify/App/blob/a7863493a03970cdc0687621ec40dc75aea99477/src/components/HTMLEngineProvider/HTMLRenderers/AnchorRenderer.js) also returns a component wrapped in a [Pressable](https://github.com/Expensify/App/blob/main/src/components/AnchorForCommentsOnly/BaseAnchorForCommentsOnly.js#L52).

**Comment by puneetlath** [comment]
@robertKozik I just wanted to check in and see how this is going.

**Comment by robertKozik** [comment]
Sorry, I was OOO most of the time. I've checked the android issue on the physical device and now it work as it should be, maybe that was a problem with my simulator. I'll check today one more just to be sure, and tomorrow PR will be ready to review, hopefully alongside the `@here` mention renderer as it is very similar

**Comment by puneetlath** [comment]
Awesome!

**Comment by puneetlath** [comment]
Nice! Reviewing asap. 

Yes, it is intended that only certain characters are allowed directly before the `@` sign in a mention and `.` is not one of them.

**Comment by robertKozik** [comment]
@puneetlath it looks like for native apps both additional padding and radius cannot be applied. The problem occurs because we are trying to style nested Text component.
I've created a snack to show it: https://snack.expo.dev/@rooobi/nested-text
More source about this behaviour from the `react-native-render-html` side:
- warning from: https://meliorence.github.io/react-native-render-html/api/renderhtmlconfig#bypassanonymoustphrasingnodes
- blog post
> padding and border are ignored in nested text elements.
 https://meliorence.github.io/react-native-render-html/blog/2021/06/29/create-blog-app-rnrh-III#frustrating-react-native-text-limitations


As it looks like we cannot properly style on native side of the things, should we remove the borderRadius and additional padding from the web also?

**Comment by puneetlath** [comment]
Just merged.

**Comment by robertKozik** [comment]
Great! I'll follow up with the here mention renderer soon

**Comment by robertKozik** [comment]
As the PR needs to be reverted due to android problems, what is the correct way to go with this? Do we pivot to the non-rounded mentions for now?

**Comment by puneetlath** [comment]
Ok I reverted [the PR](https://github.com/Expensify/App/issues/17441). This issue is the problem. 

I'm going to chat internally and see what the best path forward is here.

**Comment by robertKozik** [comment]
That catch me off guard 😢 I could have sworn that It was working for me while testing...

Let me know if you will know what next, with the work we do already here reverting to the more simple colouring will be an ease and would take no time

**Comment by puneetlath** [comment]
Ok @robertKozik @parasharrajat I chatted internally and let's go back to using styled text so it will look good on most devices and just accept that it will be ugly on Android in some scenarios. 

So back to what you had originally that had this problem: 
<img width="310" alt="Screenshot 2023-05-04 at 11 24 37 AM" src="https://user-images.githubusercontent.com/1371996/236536926-474ba0ba-1430-451b-8386-25f030a80575.png">

And then when https://github.com/Expensify/react-native/pull/43 is merged, the issue should resolve itself.

**Comment by puneetlath** [comment]
Sorry for all the flip-flopping!

**Comment by robertKozik** [comment]
Alright, It will hopefully need only reverting commits, so I'll do it right away, create new PR and test it 

> And then when https://github.com/Expensify/react-native/pull/43 is merged, the issue should resolve itself.

Not quite itself, as it will introduce new prop for such styling. But follow-up pr would need only to move style to another prop.

**Comment by puneetlath** [comment]
Sounds great and that makes sense. Thanks @robertKozik

**Comment by robertKozik** [comment]
Created draft PR for that, just want to check it on all platforms and I'll change it to `ready` cc. @parasharrajat

**Comment by robertKozik** [comment]
Sorry, forgot to change it 😄

**Comment by puneetlath** [comment]
Merged again!

**Comment by puneetlath** [comment]
All paid. Thanks everyone!
